File: packages/square/src/surface/oauth.ts
import { createHash } from 'node:crypto';
import { UnitError, json, redirect, text, type Route, type UnitContext } from '@vendor-unit/core';
import { COL, type AuthorizationCodeEntity, type MerchantEntity, type TokenEntity } from '../entities.js';
import { optionalString, readBody, requireString, type SquareDeps } from './common.js';

/**
 * OAuth surface.
 *
 * Authorize:   GET  /oauth2/authorize  https://developer.squareup.com/reference/square/oauth-api/authorize
 * ObtainToken: POST /oauth2/token      https://developer.squareup.com/reference/square/oauth-api/obtain-token
 * RevokeToken: POST /oauth2/revoke     https://developer.squareup.com/reference/square/oauth-api/revoke-token
 * TokenStatus: POST /oauth2/token/status
 *              https://developer.squareup.com/reference/square/o-auth-api/retrieve-token-status
 *
 * Documented behaviour reproduced here:
 *  - authorization codes expire after 5 minutes and are single use, and the
 *    redirect carries `code`, `response_type=code` and `state`
 *    (https://developer.squareup.com/docs/oauth-api/receive-and-manage-tokens);
 *  - a denial redirects with `error=access_denied&error_description=user_denied`
 *    (same page);
 *  - `short_lived: true` expires the access token in 24 hours, otherwise 30 days
 *    (https://developer.squareup.com/reference/square/oauth-api/obtain-token);
 *  - code-flow refresh returns the SAME refresh token, PKCE refresh returns a
 *    new single-use one that expires after 90 days
 *    (https://developer.squareup.com/docs/oauth-api/overview);
 *  - revoke takes `Authorization: Client {APPLICATION_SECRET}`, returns
 *    `{"success": true}`, and revokes every token for the merchant unless
 *    `revoke_only_access_token` is set
 *    (https://developer.squareup.com/reference/square/oauth-api/revoke-token).
 *
 * JUDGMENT — Square publishes NO error table, status codes or example error
 * bodies for /oauth2/token or /oauth2/revoke. The failures below use the
 * standard v2 error envelope (which the ObtainToken response schema does list
 * an `errors` array for) with AUTHENTICATION_ERROR/UNAUTHORIZED for credential
 * failures and INVALID_REQUEST_ERROR for malformed requests. Treat statuses on
 * these two endpoints as this mock's convention, not as Square fidelity.
 *
 * JUDGMENT — the real authorize page is an interactive consent screen. A mock
 * has nobody to click it, so approval is automatic; `unit_prompt=deny` produces
 * the documented denial redirect and `unit_prompt=html` renders a two-link
 * consent page for a human driving the flow in a browser. `unit_prompt` is the
 * only non-Square parameter in the vendor surface.
 */
export function oauthRoutes(deps: SquareDeps): Route[] {
  const { config, ids } = deps;

  return [
    {
      method: 'GET',
      path: '/oauth2/authorize',
      capability: 'oauth',
      operationId: 'Authorize',
      summary: 'Authorization page. Redirects back with an authorization code.',
      handler: ({ ctx, query }) => {
        const clientId = query('client_id');
        if (!clientId) throw new UnitError('missing_field', { detail: 'client_id is required.', field: 'client_id' });
        if (clientId !== config.applicationId) {
          throw new UnitError('invalid_value', {
            detail: `Unknown client_id '${clientId}'. This unit is configured for application '${config.applicationId}'.`,
            field: 'client_id',
          });
        }
        const redirectUri = query('redirect_uri') ?? config.redirectUri;
        if (!redirectUri) {
          throw new UnitError('missing_field', {
            detail: 'redirect_uri was not supplied and the unit has no configured redirect URL.',
            field: 'redirect_uri',
          });
        }
        const state = query('state');
        const prompt = query('unit_prompt');

        if (prompt === 'deny') {
          const url = new URL(redirectUri);
          url.searchParams.set('error', 'access_denied');
          url.searchParams.set('error_description', 'user_denied');
          if (state) url.searchParams.set('state', state);
          return redirect(url.toString());
        }

        const scopes = (query('scope') ?? config.defaultScopes.join(' ')).split(/[\s+]+/).filter(Boolean);
        const merchant = firstMerchant(ctx);

        if (prompt === 'html') {
          const approve = new URL(`http://unit.local/oauth2/authorize`);
          for (const [k, v] of Object.entries({ client_id: clientId, redirect_uri: redirectUri, scope: scopes.join(' '), ...(state ? { state } : {}) })) {
            approve.searchParams.set(k, v);
          }
          const denyUrl = new URL(approve);
          denyUrl.searchParams.set('unit_prompt', 'deny');
          return text(
            `<!doctype html><meta charset="utf-8"><title>Authorize ${merchant.businessName}</title>` +
              `<h1>${merchant.businessName}</h1><p>Grant these permissions?</p><ul>${scopes.map((s) => `<li>${s}</li>`).join('')}</ul>` +
              `<p><a href="${approve.pathname}${approve.search}">Allow</a> &middot; <a href="${denyUrl.pathname}${denyUrl.search}">Deny</a></p>`,
            200,
            { 'content-type': 'text/html; charset=utf-8' },
          );
        }

        const code = ids.authorizationCode();
        ctx.store.collection<AuthorizationCodeEntity>(COL.codes).insert(
          {
            id: code,
            clientId,
            merchantId: merchant.id,
            scopes,
            redirectUri,
            codeChallenge: query('code_challenge'),
            expiresAt: ctx.clock.isoSeconds(config.authorizationCodeTtlMs),
          },
          { operationId: 'Authorize' },
        );

        const url = new URL(redirectUri);
        url.searchParams.set('code', code);
        url.searchParams.set('response_type', 'code');
        if (state) url.searchParams.set('state', state);
        return redirect(url.toString());
      },
    },

    {
      method: 'POST',
      path: '/oauth2/token',
      capability: 'oauth',
      operationId: 'ObtainToken',
      summary: 'Exchange an authorization code, or refresh an access token.',
      handler: (args) => {
        const body = readBody(args);
        const clientId = requireString(body, 'client_id');
        const grantType = requireString(body, 'grant_type');
        if (clientId !== config.applicationId) {
          throw new UnitError('unauthorized', { detail: 'The `client_id` does not match this application.', field: 'client_id' });
        }

        if (grantType === 'authorization_code') {
          return json(exchangeCode(args.ctx, deps, body));
        }
        if (grantType === 'refresh_token') {
          return json(refreshToken(args.ctx, deps, body));
        }
        throw new UnitError('invalid_value', {
          detail: `grant_type '${grantType}' is not supported. Supported: authorization_code, refresh_token.`,
          field: 'grant_type',
          info: { supported: ['authorization_code', 'refresh_token'] },
        });
      },
    },

    {
      method: 'POST',
      path: '/oauth2/revoke',
      capability: 'oauth',
      auth: 'client-secret',
      operationId: 'RevokeToken',
      summary: 'Revoke an access token, or every token held for a merchant.',
      handler: (args) => {
        const body = readBody(args);
        const clientId = requireString(body, 'client_id');
        const accessToken = optionalString(body, 'access_token');
        const merchantId = optionalString(body, 'merchant_id');
        if (!accessToken && !merchantId) {
          throw new UnitError('missing_field', { detail: 'Provide either access_token or merchant_id.', field: 'access_token' });
        }
        if (accessToken && merchantId) {
          throw new UnitError('invalid_value', { detail: 'Do not provide access_token together with merchant_id.', field: 'merchant_id' });
        }
        const onlyThisToken = body.revoke_only_access_token === true;
        const tokens = args.ctx.store.collection<TokenEntity>(COL.tokens);

        const target = accessToken ? tokens.find((t) => t.accessToken === accessToken) : undefined;
        if (accessToken && !target) {
          throw new UnitError('unauthorized', { detail: 'The provided access token was not issued by this application.' });
        }
        const merchant = merchantId ?? target!.merchantId;
        const victims = onlyThisToken && target ? [target] : tokens.filter((t) => t.merchantId === merchant && t.clientId === clientId);
        const at = args.ctx.clock.isoMs();
        for (const v of victims) {
          if (v.revokedAt) continue;
          tokens.update(v.id, { meta: { operationId: 'RevokeToken' } }, (d) => {
            d.revokedAt = at;
          });
        }
        return json({ success: true });
      },
    },

    {
      method: 'POST',
      path: '/oauth2/token/status',
      capability: 'oauth',
      auth: 'bearer',
      operationId: 'RetrieveTokenStatus',
      summary: 'Scopes and expiry for the bearer token presented.',
      handler: ({ ctx, auth }) => {
        const token = ctx.store.collection<TokenEntity>(COL.tokens).require(auth!.tokenId!);
        return json({
          scopes: token.scopes,
          expires_at: token.expiresAt,
          client_id: token.clientId,
          merchant_id: token.merchantId,
        });
      },
    },
  ];
}

function firstMerchant(ctx: UnitContext): MerchantEntity {
  const merchants = ctx.store.collection<MerchantEntity>(COL.merchants).all();
  const merchant = merchants[0];
  if (!merchant) {
    throw new UnitError('internal', { detail: 'The seed scenario contains no merchant; OAuth cannot mint a token.' });
  }
  return merchant;
}

function exchangeCode(ctx: UnitContext, deps: SquareDeps, body: Record<string, unknown>): Record<string, unknown> {
  const { config, ids } = deps;
  const codeValue = requireString(body, 'code');
  const codes = ctx.store.collection<AuthorizationCodeEntity>(COL.codes);
  const record = codes.get(codeValue);
  if (!record) {
    throw new UnitError('unauthorized', { detail: 'The authorization code is invalid.', field: 'code' });
  }
  if (record.usedAt) {
    throw new UnitError('unauthorized', { detail: 'The authorization code has already been used. Codes are single use.', field: 'code' });
  }
  if (Date.parse(record.expiresAt) <= ctx.clock.now()) {
    throw new UnitError('unauthorized', { detail: 'The authorization code expired. Codes expire 5 minutes after they are issued.', field: 'code' });
  }

  const flow: 'code' | 'pkce' = record.codeChallenge ? 'pkce' : 'code';
  if (flow === 'pkce') {
    const verifier = requireString(body, 'code_verifier');
    const challenge = createHash('sha256').update(verifier, 'utf8').digest('base64url');
    if (challenge !== record.codeChallenge) {
      throw new UnitError('unauthorized', { detail: 'code_verifier does not match the code_challenge from the authorization request.', field: 'code_verifier' });
    }
  } else {
    const secret = requireString(body, 'client_secret');
    if (secret !== config.applicationSecret) {
      throw new UnitError('unauthorized', { detail: 'The client_secret is incorrect.', field: 'client_secret' });
    }
  }

  codes.update(record.id, { meta: { operationId: 'ObtainToken' } }, (d) => {
    d.usedAt = ctx.clock.isoMs();
  });

  const shortLived = body.short_lived === true;
  const requestedScopes = Array.isArray(body.scopes) ? (body.scopes as string[]) : record.scopes;
  return mintToken(ctx, deps, {
    clientId: record.clientId,
    merchantId: record.merchantId,
    scopes: requestedScopes,
    shortLived,
    flow,
    refreshToken: ids.refreshToken(),
  });
}

function refreshToken(ctx: UnitContext, deps: SquareDeps, body: Record<string, unknown>): Record<string, unknown> {
  const { config, ids } = deps;
  const presented = requireString(body, 'refresh_token');
  const tokens = ctx.store.collection<TokenEntity>(COL.tokens);
  const existing = tokens.find((t) => t.refreshToken === presented);
  if (!existing) {
    throw new UnitError('unauthorized', { detail: 'The refresh token is invalid.', field: 'refresh_token' });
  }
  if (existing.revokedAt) {
    throw new UnitError('unauthorized', { detail: 'The refresh token was revoked.', field: 'refresh_token' });
  }
  if (existing.flow === 'pkce') {
    if (existing.refreshTokenExpiresAt && Date.parse(existing.refreshTokenExpiresAt) <= ctx.clock.now()) {
      throw new UnitError('unauthorized', { detail: 'The refresh token expired. PKCE refresh tokens expire after 90 days.', field: 'refresh_token' });
    }
  } else {
    const secret = requireString(body, 'client_secret');
    if (secret !== config.applicationSecret) {
      throw new UnitError('unauthorized', { detail: 'The client_secret is incorrect.', field: 'client_secret' });
    }
  }

  // A refreshed code-flow token supersedes the old one; PKCE refresh tokens are
  // single use, so the old record is retired either way.
  tokens.update(existing.id, { meta: { operationId: 'ObtainToken', grant: 'refresh_token' } }, (d) => {
    d.revokedAt = ctx.clock.isoMs();
  });

  return mintToken(ctx, deps, {
    clientId: existing.clientId,
    merchantId: existing.merchantId,
    scopes: Array.isArray(body.scopes) ? (body.scopes as string[]) : existing.scopes,
    shortLived: body.short_lived === true ? true : existing.shortLived,
    flow: existing.flow,
    // Code flow keeps the same refresh token; PKCE issues a fresh one.
    refreshToken: existing.flow === 'code' ? existing.refreshToken : ids.refreshToken(),
  });
}

function mintToken(
  ctx: UnitContext,
  deps: SquareDeps,
  spec: { clientId: string; merchantId: string; scopes: string[]; shortLived: boolean; flow: 'code' | 'pkce'; refreshToken: string },
): Record<string, unknown> {
  const { config, ids } = deps;
  const ttl = spec.shortLived ? config.shortLivedTtlMs : config.accessTokenTtlMs;
  const accessToken = ids.accessToken();
  const expiresAt = ctx.clock.isoSeconds(ttl);
  const refreshExpiresAt = spec.flow === 'pkce' ? ctx.clock.isoSeconds(config.pkceRefreshTtlMs) : undefined;

  ctx.store.collection<TokenEntity>(COL.tokens).insert(
    {
      id: ids.internal('tok'),
      accessToken,
      refreshToken: spec.refreshToken,
      clientId: spec.clientId,
      merchantId: spec.merchantId,
      scopes: spec.scopes,
      expiresAt,
      refreshTokenExpiresAt: refreshExpiresAt,
      shortLived: spec.shortLived,
      flow: spec.flow,
    },
    { operationId: 'ObtainToken' },
  );

  return {
    access_token: accessToken,
    token_type: 'bearer',
    expires_at: expiresAt,
    merchant_id: spec.merchantId,
    refresh_token: spec.refreshToken,
    short_lived: spec.shortLived,
    ...(refreshExpiresAt ? { refresh_token_expires_at: refreshExpiresAt } : {}),
  };
}

File: src/vendorfake/clover/surface/oauth.py
"""The OAuth v2 surface: authorize, token exchange, refresh.

FOR: reproducing the three endpoints a Clover OAuth consumer drives, with the
documented redirect parameters, the documented four-field token response and
the documented single-use refresh rotation -- so that an integration which
handles rotation correctly against this unit handles it correctly against
Clover, and one which does not fails here first.

=========  =================================================================
Authorize  ``GET  /oauth/v2/authorize``
           https://docs.clover.com/dev/docs/high-trust-app-auth-flow
Token      ``POST /oauth/v2/token``
           https://docs.clover.com/dev/docs/generate-oauth-expiring-access-and-refresh-token
Refresh    ``POST /oauth/v2/refresh``
           https://docs.clover.com/dev/docs/refresh-access-tokens
=========  =================================================================

Documented behaviour reproduced here
------------------------------------
* the authorize redirect carries ``merchant_id``, ``client_id`` and ``code``
  -- Clover, unlike Square, identifies the merchant and echoes the app in the
  callback (high-trust-app-auth-flow, verbatim shape);
* token exchange accepts ``{client_id, client_secret, code}`` (high-trust) or
  ``{client_id, code, code_verifier}`` (PKCE) and answers exactly
  ``{access_token, access_token_expiration, refresh_token,
  refresh_token_expiration}``, expirations in Unix **seconds**;
* refresh takes ``{client_id, refresh_token}`` with **no client_secret**, and
  "Refresh token is for single use and becomes invalid immediately after a
  new access_token and refresh_token pair is generated" -- reusing a rotated
  refresh token is refused;
* access tokens live 30 minutes ("OAuth access_tokens expire in 30 minutes",
  oauth-and-tokens-faqs).

JUDGMENT, each labelled at its site
-----------------------------------
* **Error bodies.** Clover documents no OAuth error body, status or code
  anywhere (audit gap 2). Failures answer the package envelope
  (``{"message": ...}``) with 401 for credential failures and 400 for
  malformed requests; the bad-code message is the one phrase Clover's own FAQ
  uses, "Failed to validate authentication code"
  (https://docs.clover.com/dev/docs/oauth-and-tokens-faqs), answered
  identically for an unknown, spent or expired code -- a real authorization
  server does not say which, and the ``unit_error`` sidecar carries the
  distinction for whoever is debugging.
* **Code TTL and single use** -- documented nowhere; ten minutes and
  single-use per RFC 6749 convention, from ``CloverConfig``.
* **``state``** -- Clover's v2 authorize example shows no ``state`` parameter.
  It is passed through to the redirect when the consumer sends one, because
  every standard OAuth client library sends and verifies it and a fake that
  dropped it would break CSRF checks the consumer is right to run. A consumer
  must not conclude from this unit that Clover itself echoes it.
* **PKCE at authorize** -- the *token* side of PKCE is documented
  (``code_verifier`` in the exchange body); the authorize-side
  ``code_challenge`` parameter is not shown on the v2 pages, so accepting it
  here follows RFC 7636, S256 only -- and an omitted ``code_challenge_method``
  is refused rather than defaulted, because the RFC's default is ``plain``.
  A code issued with a challenge demands the verifier at exchange; one issued
  without demands the secret.
* **Prior access tokens survive a refresh.** Clover's docs invalidate only
  the *refresh* token on rotation and say nothing about access tokens minted
  earlier, so this unit keeps them valid to their own expiry. Inventing
  revocation would teach consumers an invalidation rule Clover does not
  publish, and they would discover its absence in production.

THE ORDERING INVARIANT (the N-1 class from the Square build): **no 4xx leaves
a journal entry.** Every refusal on the exchange path -- unknown client, bad
code, expired code, secret, verifier -- is computed before the code's
mark-used write, and every refusal on the refresh path before the rotation
write. A refused request must leave the world exactly as it found it, so the
consumer's corrected retry with the same code or refresh token succeeds.
"""

from __future__ import annotations

import hashlib
from collections.abc import Mapping
from typing import Any
from urllib.parse import parse_qsl, urlencode, urlsplit, urlunsplit

from vendorfake.clover.entities import COL, AuthorizationCodeEntity, MerchantEntity, TokenEntity
from vendorfake.clover.model.common import validate_body
from vendorfake.clover.model.oauth import RefreshRequest, TokenRequest, TokenResponse
from vendorfake.clover.surface.common import CloverDeps, is_past_ms, wire_seconds
from vendorfake.core.kernel.reply import json_, redirect
from vendorfake.core.kernel.types import (
    HandlerArgs,
    ReplyInit,
    Route,
    UnitContext,
    UnitError,
    UnitErrorKind,
)
from vendorfake.core.state.store import Entity
from vendorfake.core.util.b64 import b64url_encode

__all__ = ["CAPABILITY", "FAILED_CODE_MESSAGE", "CloverOAuthSurface", "oauth_routes"]

CAPABILITY = "oauth"
"""The capability every route below belongs to."""

FAILED_CODE_MESSAGE = "Failed to validate authentication code"
"""Clover's own phrase for a code it will not accept, from the OAuth FAQ
(https://docs.clover.com/dev/docs/oauth-and-tokens-faqs). Used verbatim for an
unknown, spent and expired code alike; see the module docstring."""


class CloverOAuthSurface:
    """The three OAuth routes, bound to one vendor's config and id stream."""

    __slots__ = ("_deps",)

    def __init__(self, deps: CloverDeps) -> None:
        self._deps = deps

    def routes(self) -> tuple[Route, ...]:
        return (
            Route(
                method="GET",
                path="/oauth/v2/authorize",
                capability=CAPABILITY,
                handler=self.authorize,
                operation_id="Authorize",
                summary="Authorization page. Redirects back with merchant_id, client_id and a code.",
            ),
            Route(
                method="POST",
                path="/oauth/v2/token",
                capability=CAPABILITY,
                handler=self.exchange_token,
                operation_id="ExchangeToken",
                summary="Exchange an authorization code for the four-field token response.",
            ),
            Route(
                method="POST",
                path="/oauth/v2/refresh",
                capability=CAPABILITY,
                handler=self.refresh_token,
                operation_id="RefreshToken",
                summary="Rotate a single-use refresh token for a new access/refresh pair.",
            ),
        )

    # -- GET /oauth/v2/authorize -------------------------------------------

    def authorize(self, args: HandlerArgs) -> ReplyInit:
        """The consent redirect, approval automatic (a mock has nobody to click).

        The redirect query is the documented shape:
        ``?merchant_id=...&client_id=...&code=...``
        (https://docs.clover.com/dev/docs/high-trust-app-auth-flow).
        """
        config = self._deps.config
        client_id = args.query("client_id")
        if not client_id:
            raise UnitError(UnitErrorKind.MISSING_FIELD, detail="client_id is required.", field="client_id")
        if client_id != config.client_id:
            raise UnitError(
                UnitErrorKind.INVALID_VALUE,
                detail=f"Unknown client_id {client_id!r}. This unit is configured for app {config.client_id!r}.",
                field="client_id",
            )

        supplied_redirect = args.query("redirect_uri")
        if supplied_redirect and supplied_redirect != config.redirect_uri:
            # The mismatch this parameter exists to catch. JUDGMENT on the
            # status -- Clover documents no error for it -- but not on the
            # refusal: redirecting a code to an unregistered URI is the attack.
            raise UnitError(
                UnitErrorKind.INVALID_VALUE,
                detail="redirect_uri does not match the redirect URI registered for this app.",
                field="redirect_uri",
            )
        redirect_uri = supplied_redirect or config.redirect_uri

        challenge = args.query("code_challenge")
        method = args.query("code_challenge_method")
        if challenge and method != "S256":
            # JUDGMENT: an omitted method is refused, not defaulted to S256.
            # RFC 7636 s4.3 makes `plain` the default when the method is
            # absent, and this unit refuses `plain`; silently upgrading an
            # absent method to S256 would accept a request the RFC reads as
            # the one method we reject. Clover documents nothing either way.
            raise UnitError(
                UnitErrorKind.INVALID_VALUE,
                detail=(
                    "code_challenge_method must be S256. An omitted method means 'plain' (RFC 7636 s4.3), "
                    "which is not supported."
                ),
                field="code_challenge_method",
            )

        merchant = _the_merchant(args.ctx)
        code = self._deps.ids.authorization_code()
        args.ctx.store.collection(COL.codes).insert(
            AuthorizationCodeEntity(
                id=code,
                client_id=client_id,
                merchant_id=merchant.id,
                expires_at_ms=int(args.ctx.clock.now()) + self._deps.config.authorization_code_ttl_ms,
                code_challenge=challenge,
            ).to_entity(),
            {"operation_id": "Authorize"},
        )
        params = {"merchant_id": merchant.id, "client_id": client_id, "code": code}
        state = args.query("state")
        if state is not None:
            # JUDGMENT: not in Clover's documented redirect; see the module
            # docstring for why it is echoed anyway.
            params["state"] = state
        return redirect(_with_query(redirect_uri, params))

    # -- POST /oauth/v2/token ----------------------------------------------

    def exchange_token(self, args: HandlerArgs) -> ReplyInit:
        ctx = args.ctx
        request = validate_body(TokenRequest, args.body())
        self._check_client(request.client_id)

        codes = ctx.store.collection(COL.codes)
        stored = codes.get(request.code)
        if stored is None:
            raise _failed_code("unknown")
        record = AuthorizationCodeEntity.from_entity(stored)
        if record.used_at_ms is not None:
            raise _failed_code("already_used")
        if is_past_ms(record.expires_at_ms, ctx.clock):
            raise _failed_code("expired")

        # Every refusal above and below this comment happens BEFORE the
        # mark-used write: a refused exchange must not burn the code (the
        # module docstring's ordering invariant).
        if record.code_challenge is not None:
            self._check_verifier(request.code_verifier, record.code_challenge)
        else:
            self._check_secret(request.client_secret)

        now_ms = int(ctx.clock.now())

        def mark_used(draft: Entity) -> None:
            draft["used_at_ms"] = now_ms

        codes.update(record.id, mark_used, meta={"operation_id": "ExchangeToken"})
        return json_(
            self._mint(
                ctx,
                client_id=record.client_id,
                merchant_id=record.merchant_id,
                now_ms=now_ms,
                operation_id="ExchangeToken",
            )
        )

    # -- POST /oauth/v2/refresh --------------------------------------------

    def refresh_token(self, args: HandlerArgs) -> ReplyInit:
        ctx = args.ctx
        request = validate_body(RefreshRequest, args.body())
        self._check_client(request.client_id)

        tokens = ctx.store.collection(COL.tokens)
        found = tokens.find(lambda entity: entity.get("refresh_token") == request.refresh_token)
        if found is None:
            raise UnitError(
                UnitErrorKind.UNAUTHORIZED,
                detail="The refresh token is invalid.",
                field="refresh_token",
            )
        existing = TokenEntity.from_entity(found)
        if existing.client_id != request.client_id:
            # Unreachable while one app is configured (`_check_client` already
            # matched the request to it), but a second seeded app must not be
            # able to rotate the first app's grant. Same 401 as any other
            # credential failure, and above every write.
            raise UnitError(
                UnitErrorKind.UNAUTHORIZED,
                detail="The refresh token was not issued to this client_id.",
                field="refresh_token",
            )
        if existing.refresh_used_at_ms is not None:
            # "Refresh token is for single use and becomes invalid immediately
            # after a new access_token and refresh_token pair is generated."
            # https://docs.clover.com/dev/docs/refresh-access-tokens
            raise UnitError(
                UnitErrorKind.UNAUTHORIZED,
                detail="The refresh token was already used. Refresh tokens are single use.",
                field="refresh_token",
            )
        if is_past_ms(existing.refresh_token_expiration_ms, ctx.clock):
            raise UnitError(
                UnitErrorKind.UNAUTHORIZED,
                detail="The refresh token expired.",
                field="refresh_token",
            )

        # All refusals are above; the rotation write is below. A refused
        # refresh must not retire the token -- with single-use rotation that
        # would lock the consumer out of the grant permanently, answered by a
        # 4xx claiming nothing happened.
        now_ms = int(ctx.clock.now())

        def rotate(draft: Entity) -> None:
            draft["refresh_used_at_ms"] = now_ms

        # Journalled: rotation is a real, documented state change a consumer
        # can observe (the old refresh token stops working). The old ACCESS
        # token is deliberately untouched; see the module docstring.
        tokens.update(existing.id, rotate, meta={"operation_id": "RefreshToken"})
        return json_(
            self._mint(
                ctx,
                client_id=existing.client_id,
                merchant_id=existing.merchant_id,
                now_ms=now_ms,
                operation_id="RefreshToken",
            )
        )

    # -- shared ------------------------------------------------------------

    def _check_client(self, client_id: str) -> None:
        """JUDGMENT on the status: no OAuth error is documented, and 401 is
        the package convention for a credential failure."""
        if client_id != self._deps.config.client_id:
            raise UnitError(
                UnitErrorKind.UNAUTHORIZED,
                detail="The client_id does not match this app.",
                field="client_id",
            )

    def _check_secret(self, presented: str | None) -> None:
        """The high-trust flow authenticates with the app secret in the body."""
        if not presented:
            raise UnitError(
                UnitErrorKind.MISSING_FIELD,
                detail="client_secret is required for a code issued without a code_challenge.",
                field="client_secret",
            )
        if presented != self._deps.config.client_secret:
            raise UnitError(
                UnitErrorKind.UNAUTHORIZED,
                detail="The client_secret is incorrect.",
                field="client_secret",
            )

    def _check_verifier(self, verifier: str | None, challenge: str) -> None:
        """PKCE proves knowledge of the verifier: S256 only,
        ``BASE64URL(SHA256(ASCII(code_verifier)))`` unpadded (RFC 7636)."""
        if not verifier:
            raise UnitError(
                UnitErrorKind.MISSING_FIELD,
                detail="code_verifier is required for a code issued with a code_challenge.",
                field="code_verifier",
            )
        computed = b64url_encode(hashlib.sha256(verifier.encode("utf-8")).digest())
        if computed != challenge:
            raise UnitError(
                UnitErrorKind.UNAUTHORIZED,
                detail="code_verifier does not match the code_challenge from the authorization request.",
                field="code_verifier",
            )

    def _mint(
        self,
        ctx: UnitContext,
        *,
        client_id: str,
        merchant_id: str,
        now_ms: int,
        operation_id: str,
    ) -> dict[str, Any]:
        """Issue one access/refresh pair, record it, and answer the documented
        four fields. Stored instants are ms; the wire gets Unix seconds
        through :func:`~vendorfake.clover.surface.common.wire_seconds`.

        ``operation_id`` is the caller's, so a refresh journals the token it
        minted as a refresh: one request, one operation, however many writes.
        """
        config = self._deps.config
        access_expires_ms = now_ms + config.access_token_ttl_ms
        refresh_expires_ms = now_ms + config.refresh_token_ttl_ms
        entity = TokenEntity(
            id=self._deps.ids.internal("tok"),
            access_token=self._deps.ids.access_token(),
            refresh_token=self._deps.ids.refresh_token(),
            client_id=client_id,
            merchant_id=merchant_id,
            access_token_expiration_ms=access_expires_ms,
            refresh_token_expiration_ms=refresh_expires_ms,
            # Every token inherits the app's fixed permission set; Clover has
            # no per-token scope request. See config.py.
            permissions=config.permissions,
            createdTime=now_ms,
        )
        ctx.store.collection(COL.tokens).insert(entity.to_entity(), {"operation_id": operation_id})
        return TokenResponse(
            access_token=entity.access_token,
            access_token_expiration=wire_seconds(access_expires_ms),
            refresh_token=entity.refresh_token,
            refresh_token_expiration=wire_seconds(refresh_expires_ms),
        ).wire()


# ---------------------------------------------------------------------------
# Module-level helpers: pure, and testable without a unit.
# ---------------------------------------------------------------------------


def oauth_routes(deps: CloverDeps) -> tuple[Route, ...]:
    """The OAuth routes for one vendor."""
    return CloverOAuthSurface(deps).routes()


def _failed_code(reason: str) -> UnitError:
    """The one documented phrase for every kind of bad code; the sidecar says
    which kind, for whoever is debugging. See the module docstring."""
    return UnitError(
        UnitErrorKind.UNAUTHORIZED,
        detail=FAILED_CODE_MESSAGE,
        field="code",
        info={"reason": reason},
    )


def _the_merchant(ctx: UnitContext) -> MerchantEntity:
    """The merchant this unit represents.

    One per unit; the first insertion-ordered record is the answer. Until the
    PR-E seed ships one, a unit doing OAuth needs a merchant inserted by its
    driver, and the error says so.
    """
    merchants = ctx.store.collection(COL.merchants).all()
    if not merchants:
        raise UnitError(
            UnitErrorKind.INTERNAL,
            detail="This unit has no merchant; OAuth cannot mint a code. The seed scenario arrives in PR E.",
        )
    return MerchantEntity.from_entity(merchants[0])


def _with_query(url: str, params: Mapping[str, str]) -> str:
    """``url`` with ``params`` set on its query string.

    Built with :func:`urllib.parse.urlencode` rather than by concatenation, so
    a redirect URI that already carries a query keeps it and a ``state`` value
    containing ``&`` survives the round trip. Existing occurrences of a key
    are replaced rather than appended.
    """
    parts = urlsplit(url)
    kept = [(name, value) for name, value in parse_qsl(parts.query, keep_blank_values=True) if name not in params]
    query = urlencode([*kept, *params.items()])
    return urlunsplit((parts.scheme, parts.netloc, parts.path, query, parts.fragment))
